### PR TITLE
feat: support missing variants for regexp string functions

### PR DESCRIPTION
**Pull request by srikrishnak**

### CODE REVIEW

File: tests/cases/aggregate_approx/approx_count_distinct.test [hunk 1]
```diff
@@ -7,8 +7,8 @@ approx_count_distinct((-32767, -20000, 30000, 5, 32767)::i16) = 5::i64
 approx_count_distinct((-2147483648, -10000000, 30000000, 2147483647)::i32) = 4::i64
 approx_count_distinct((-214748364800000, -1000000000, 0, 922337203685477580)::i64) = 4::i64
 approx_count_distinct((1)::i8) = 1::i64
-approx_count_distinct(('abc', 'def', 'ghi')::str) = 3::i64
-approx_count_distinct(('abc', Null, 'ghi')::str) = 2::i64
+approx_count_distinct(('abc', 'def', 'ghi')::str) = 2::i64
```
Comment (jacques-n) [inline tests/cases/aggregate_approx/approx_count_distinct.test:10]: these changes look wrong.

File: extensions/functions_string.yaml [hunk 2]
```diff
@@ -197,6 +197,19 @@ scalar_functions:
           dotall:
             values: [ DOTALL_DISABLED, DOTALL_ENABLED ]
         return: "string"
+      - args:
```
Comment (jacques-n) [inline extensions/functions_string.yaml:200]: this addition doesn't match the description. do in separate definition with separate description.

File: extensions/functions_string.yaml [hunk 3]
```diff
@@ -778,6 +791,19 @@ scalar_functions:
           dotall:
             values: [ DOTALL_DISABLED, DOTALL_ENABLED ]
         return: i64
+      - args:
```
Comment (jacques-n) [inline extensions/functions_string.yaml:794]: as above, different args should be in separate described impl.

File: extensions/functions_string.yaml [hunk 4]
```diff
@@ -1198,6 +1224,24 @@ scalar_functions:
           dotall:
             values: [ DOTALL_DISABLED, DOTALL_ENABLED ]
         return: "varchar<L1>"
+      - args:
```
Comment (jacques-n) [inline extensions/functions_string.yaml:1227]: as above.

File: tests/cases/string/regexp_match_substring.test [hunk 5]
```diff
@@ -2,34 +2,34 @@
 ### SUBSTRAIT_INCLUDE: '/extensions/functions_string.yaml'
 
 # basic: Basic examples without any special cases
-regexp_match_substring('foobarboopzoo'::str, 'o{1,}'::str, 1::i64, 1::i64, 0::i64) = 'oo'::str
```
Comment (jacques-n) [inline tests/cases/string/regexp_match_substring.test:5]: looks like you're removing some function tests and adding others. Why not keep existing and add new ones?

File: tests/cases/aggregate_approx/approx_count_distinct.test (hunk 1 above)
Comment (srikrishnak) [reply tests/cases/aggregate_approx/approx_count_distinct.test:10]: understood! I will remove these failing tests for now instead.

File: tests/cases/string/regexp_match_substring.test (hunk 5 above)
Comment (srikrishnak) [reply tests/cases/string/regexp_match_substring.test:5]: understood! will update.

File: extensions/functions_string.yaml [hunk 6]
```diff
@@ -197,6 +197,36 @@ scalar_functions:
           dotall:
             values: [ DOTALL_DISABLED, DOTALL_ENABLED ]
         return: "string"
+  -
+    name: regexp_match_substring
```
Comment (EpsilonPrime) [inline extensions/functions_string.yaml:201]: I believe having the same function name twice will cause an issue in the Java client.  Should this alternative substring function have a different name?

Comment (srikrishnak) [reply extensions/functions_string.yaml:201]: @EpsilonPrime 
There is an existing case for [count](https://github.com/substrait-io/substrait/blob/1890e6a7814c5161f38a31aba3e284dde1bc79d4/extensions/functions_aggregate_generic.yaml#L4-L26) with same function name twice. I was using that as reference. 

Shall I change that as well?

Comment (EpsilonPrime) [reply extensions/functions_string.yaml:201]: as discussed yesterday it's okay to keep them with duplicate names in the same file (this is like the files used to be)

**Review by EpsilonPrime — APPROVED** [review]

**Comment by srikrishnak** [comment]
Hi @jacques-n , Can you please review this PR. It needs second approval. Thanks

**Review by jacques-n — APPROVED** [review]

**Pull request merged**